core/io/remote_filesystem_client.cpp

```diff
@@ -0,0 +1,333 @@
+		String l = p_cache[i].path + "::" + itos(p_cache[i].server_modified_time) + "::" + itos(p_cache[i].modified_time);
+		f->store_line(l);
+	}
+	return OK;
+}
+
+Error RemoteFilesystemClient::synchronize_with_server(const String &p_host, int p_port, const String &p_password, String &r_cache_path) {
+	Error err = _synchronize_with_server(p_host, p_port, p_password, r_cache_path);
+	// Ensure no memory is kept
+	validated_directories.reset();
+	cache_path = String();
+	return err;
+}
+
+void RemoteFilesystemClient::_update_cache_path(String &r_cache_path) {
+	r_cache_path = cache_path.path_join(FILES_SUBFOLDER);
+}
+
+Error RemoteFilesystemClient::_synchronize_with_server(const String &p_host, int p_port, const String &p_password, String &r_cache_path) {
+	cache_path = r_cache_path;
+	{
+		Ref<DirAccess> dir = DirAccess::create(DirAccess::ACCESS_FILESYSTEM);
+		dir->change_dir(cache_path);
+		cache_path = dir->get_current_dir();
+	}
+
+	Ref<StreamPeerTCP> tcp_client;
+	tcp_client.instantiate();
+
+	IPAddress ip = p_host.is_valid_ip_address() ? IPAddress(p_host) : IP::get_singleton()->resolve_hostname(p_host);
+	ERR_FAIL_COND_V_MSG(!ip.is_valid(), ERR_INVALID_PARAMETER, "Unable to resolve remote filesystem server hostname: " + p_host);
+	print_verbose("Remote Filesystem: Connecting to host " + String(ip) + ", port: " + itos(p_port));
+	Error err = tcp_client->connect_to_host(ip, p_port);
+	ERR_FAIL_COND_V_MSG(err != OK, err, "Unable to open connection to remote file server (" + String(p_host) + ", port " + itos(p_port) + ") failed.");
+
+	while (tcp_client->get_status() == StreamPeerTCP::STATUS_CONNECTING) {
+		tcp_client->poll();
+		OS::get_singleton()->delay_usec(100);
+	}
+
+	if (tcp_client->get_status() != StreamPeerTCP::STATUS_CONNECTED) {
+		ERR_FAIL_V_MSG(ERR_CANT_CONNECT, "Connection to remote file server (" + String(p_host) + ", port " + itos(p_port) + ") failed.");
+	}
+
+	// Connection OK, now send the current file state.
+	print_verbose("Remote Filesystem: Connection OK.");
+
+	// Header (GRFS) - Godot Remote File System
+	print_verbose("Remote Filesystem: Sending Header");
+	tcp_client->put_u8('G');
+	tcp_client->put_u8('R');
+	tcp_client->put_u8('F');
+	tcp_client->put_u8('S');
+	// Protocol version
+	tcp_client->put_32(FILESYSTEM_PROTOCOL_VERSION);
+	print_verbose("Remote Filesystem: Send password");
+	uint8_t password[PASSWORD_LENGTH]; // Send fixed size password, since its easier and safe to validate.
+	for (int i = 0; i < PASSWORD_LENGTH; i++) {
+		if (i < p_password.length()) {
+			password[i] = p_password[i];
+		} else {
+			password[i] = 0;
+		}
+	}
+	tcp_client->put_data(password, PASSWORD_LENGTH);
+	print_verbose("Remote Filesystem: Tags");
+	Vector<String> tags;
+	{
+		String os_name = OS::get_singleton()->get_name().to_lower();
+		if (os_name == "Linux" || os_name == "FreeBSD" || os_name == "NetBSD" || os_name == "OpenBSD" || os_name == "BSD") {
+			os_name = "linuxbsd";
+		}
+		tags.push_back(os_name);
```

Comment:
What's the purpose of this?
Also, none of the literal strings will match the lowered OS name.
If it's needed to know if the system is POSIX, UNIX-like or so, maybe there's a more robust check to do.

Reply:
This is harcoded on OS_Unix so I kind of need to replicate the test, but it makes sense I will redo the code so its more robust.